File: RandomEvent.cpp
#include "RandomEvent.h"
// НЕ НУЖНО включать VirtualPet.h - избегаем циклической зависимости

RandomEvent::RandomEvent() : eventsTriggered(0) {
    // ПОЛОЖИТЕЛЬНЫЕ СОБЫТИЯ
    events.push_back({
        "Щедрый сосед",
        "Сосед угостил питомца вкусняшками!",
        -30, 0, 5, 15, 10
        });
    events.push_back({
        "Находка в парке",
        "Питомец нашёл потерянную игрушку!",
        0, -10, 0, 20, 15
        });
    events.push_back({
        "Солнечный день",
        "Хорошая погода подняла настроение!",
        0, -15, 5, 10, 5
        });
    events.push_back({
        "Подарок от друга",
        "Друг прислал посылку с кормом!",
        -40, 0, 0, 10, 0
        });
    events.push_back({
        "Удачная тренировка",
        "Питомец освоил новый трюк!",
        -10, -20, 0, 15, 25
        });
    events.push_back({
        "Лотерея выигрыш",
        "Вы выиграли 500 монет в лотерею!",
        0, 0, 0, 25, 30
        });
    events.push_back({
        "Встреча с чемпионом",
        "Известный тренер похвалил питомца!",
        0, 0, 5, 20, 20
        });
    events.push_back({
        "Магазин скидок",
        "Скидка 50% на все товары!",
        -20, 0, 0, 10, 0
        });
    events.push_back({
        "Новый друг",
        "Питомец нашёл нового друга!",
        0, -5, 0, 30, 15
        });
    events.push_back({
        "День рождения",
        "У питомца день рождения!",
        -25, -10, 10, 40, 50
        });

    // ОТРИЦАТЕЛЬНЫЕ СОБЫТИЯ
    events.push_back({
        "Дождливая погода",
        "Плохая погода испортила настроение",
        5, 10, -5, -15, 0
        });
    events.push_back({
        "Потерянная игрушка",
        "Любимая игрушка потерялась",
        0, 5, 0, -20, 0
        });
    events.push_back({
        "Визит к ветеринару",
        "Плановый осмотр у врача",
        0, -30, 5, -10, 10
        });
    events.push_back({
        "Голодный день",
        "Корм закончился раньше времени",
        25, 5, -5, -15, 0
        });
    events.push_back({
        "Бессонная ночь",
        "Питомец плохо спал ночью",
        5, 25, -5, -10, 0
        });
    events.push_back({
        "Ссора с соседом",
        "Соседский кот украл корм",
        20, 0, 0, -25, 0
        });
    events.push_back({
        "Простуда",
        "Питомец немного простудился",
        5, 10, -15, -10, 0
        });
    events.push_back({
        "Скучный день",
        "Ничего интересного не произошло",
        5, 5, 0, -10, 0
        });

    // НЕЙТРАЛЬНЫЕ СОБЫТИЯ
    events.push_back({
        "Обычный день",
        "Ничего особенного не случилось",
        5, 5, 0, 0, 5
        });
    events.push_back({
        "Прогулка",
        "Обычная прогулка во дворе",
        -5, -10, 0, 5, 10
        });
    events.push_back({
        "Дремота",
        "Питомец дремал весь день",
        5, -20, 0, 0, 5
        });
}

bool RandomEvent::trigger(PetBase& pet) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> chanceDist(1, 100);

    int chance = chanceDist(gen);
    if (chance > 30) {
        return false;
    }

    std::uniform_int_distribution<> eventDist(0, events.size() - 1);
    int eventIndex = eventDist(gen);
    const GameEvent& event = events[eventIndex];

    pet.getParameters().setHunger(
        pet.getParameters().getHunger() + event.hungerChange
    );
    pet.getParameters().setFatigue(
        pet.getParameters().getFatigue() + event.fatigueChange
    );
    pet.getParameters().setHealth(
        pet.getParameters().getHealth() + event.healthChange
    );
    pet.getParameters().setMood(
        pet.getParameters().getMood() + event.moodChange
    );

    // ИСПРАВЛЕНО: просто вызываем виртуальный метод
    if (event.xpChange > 0) {
        pet.addXP(event.xpChange);  // Работает для всех типов питомцев
    }

    std::cout << "\n";
    std::cout << "************************************************\n";
    std::cout << "*              СЛУЧАЙНОЕ СОБЫТИЕ!              *\n";
    std::cout << "************************************************\n";
    std::cout << "* " << event.name << "\n";
    std::cout << "*----------------------------------------------*\n";
    std::cout << "* " << event.description << "\n";

    if (event.hungerChange != 0) {
        std::cout << "* Голод: " << (event.hungerChange > 0 ? "+" : "")
            << event.hungerChange << "\n";
    }
    if (event.fatigueChange != 0) {
        std::cout << "* Усталость: " << (event.fatigueChange > 0 ? "+" : "")
            << event.fatigueChange << "\n";
    }
    if (event.healthChange != 0) {
        std::cout << "* Здоровье: " << (event.healthChange > 0 ? "+" : "")
            << event.healthChange << "\n";
    }
    if (event.moodChange != 0) {
        std::cout << "* Настроение: " << (event.moodChange > 0 ? "+" : "")
            << event.moodChange << "\n";
    }
    if (event.xpChange > 0) {
        std::cout << "* Опыт: +" << event.xpChange << "\n";
    }

    std::cout << "************************************************\n";

    eventsTriggered++;
    return true;
}

void RandomEvent::showEventLog() const {
    std::cout << "\n************************************************\n";
    std::cout << "*           Журнал событий                     *\n";
    std::cout << "************************************************\n";
    std::cout << "* Всего событий произошло: " << eventsTriggered << "              *\n";
    std::cout << "************************************************\n";
}

File: RandomEvent.cs
using System;
using System.Collections.Generic;

public class GameEvent
{
    public string Name { get; set; }
    public string Description { get; set; }
    public int HungerChange { get; set; }
    public int FatigueChange { get; set; }
    public int HealthChange { get; set; }
    public int MoodChange { get; set; }
    public int XPChange { get; set; }
}

public class RandomEvent
{
    private List<GameEvent> events;
    private int eventsTriggered;

    public RandomEvent()
    {
        events = new List<GameEvent>();
        eventsTriggered = 0;
        InitializeEvents();
    }

    private void InitializeEvents()
    {
        // Положительные события
        events.Add(new GameEvent { Name = "Щедрый сосед", Description = "Сосед угостил питомца вкусняшками!", HungerChange = -30, FatigueChange = 0, HealthChange = 5, MoodChange = 15, XPChange = 10 });
        events.Add(new GameEvent { Name = "Находка в парке", Description = "Питомец нашёл потерянную игрушку!", HungerChange = 0, FatigueChange = -10, HealthChange = 0, MoodChange = 20, XPChange = 15 });
        events.Add(new GameEvent { Name = "Лотерея выигрыш", Description = "Вы выиграли 500 монет в лотерею!", HungerChange = 0, FatigueChange = 0, HealthChange = 0, MoodChange = 25, XPChange = 30 });
        events.Add(new GameEvent { Name = "День рождения", Description = "У питомца день рождения!", HungerChange = -25, FatigueChange = -10, HealthChange = 10, MoodChange = 40, XPChange = 50 });

        // Отрицательные события
        events.Add(new GameEvent { Name = "Дождливая погода", Description = "Плохая погода испортила настроение", HungerChange = 5, FatigueChange = 10, HealthChange = -5, MoodChange = -15, XPChange = 0 });
        events.Add(new GameEvent { Name = "Простуда", Description = "Питомец немного простудился", HungerChange = 5, FatigueChange = 10, HealthChange = -15, MoodChange = -10, XPChange = 0 });
        events.Add(new GameEvent { Name = "Голодный день", Description = "Корм закончился раньше времени", HungerChange = 25, FatigueChange = 5, HealthChange = -5, MoodChange = -15, XPChange = 0 });
    }

    public bool Trigger(PetBase pet)
    {
        Random random = new Random();
        int chance = random.Next(1, 101);

        if (chance > 30)
            return false;

        int eventIndex = random.Next(0, events.Count);
        var gameEvent = events[eventIndex];

        pet.GetParameters().SetHunger(pet.GetParameters().GetHunger() + gameEvent.HungerChange);
        pet.GetParameters().SetFatigue(pet.GetParameters().GetFatigue() + gameEvent.FatigueChange);
        pet.GetParameters().SetHealth(pet.GetParameters().GetHealth() + gameEvent.HealthChange);
        pet.GetParameters().SetMood(pet.GetParameters().GetMood() + gameEvent.MoodChange);

        if (gameEvent.XPChange > 0 && pet is VirtualPet vPet)
        {
            vPet.AddXP(gameEvent.XPChange);
        }

        Console.WriteLine("\n╔════════════════════════════════════════════════════════╗");
        Console.WriteLine("║              СЛУЧАЙНОЕ СОБЫТИЕ!                        ║");
        Console.WriteLine("╠════════════════════════════════════════════════════════╣");
        Console.WriteLine($"║ {gameEvent.Name,-56} ║");
        Console.WriteLine("╟────────────────────────────────────────────────────────╢");
        Console.WriteLine($"║ {gameEvent.Description,-56} ");

        if (gameEvent.HungerChange != 0)
            Console.WriteLine($"║ Голод: {(gameEvent.HungerChange > 0 ? "+" : "")}{gameEvent.HungerChange,-3}                                        ");
        if (gameEvent.FatigueChange != 0)
            Console.WriteLine($"║ Усталость: {(gameEvent.FatigueChange > 0 ? "+" : "")}{gameEvent.FatigueChange,-3}                                      ");
        if (gameEvent.HealthChange != 0)
            Console.WriteLine($"║ Здоровье: {(gameEvent.HealthChange > 0 ? "+" : "")}{gameEvent.HealthChange,-3}                                       ");
        if (gameEvent.MoodChange != 0)
            Console.WriteLine($"║ Настроение: {(gameEvent.MoodChange > 0 ? "+" : "")}{gameEvent.MoodChange,-3}                                       ");
        if (gameEvent.XPChange > 0)
            Console.WriteLine($"║ Опыт: +{gameEvent.XPChange,-2}                                             ");

        Console.WriteLine("╚════════════════════════════════════════════════════════╝");

        eventsTriggered++;
        return true;
    }

    public void ShowEventLog()
    {
        Console.WriteLine("\n╔════════════════════════════════════════════════════════╗");
        Console.WriteLine("║           Журнал событий                               ");
        Console.WriteLine("╠════════════════════════════════════════════════════════╣");
        Console.WriteLine($"║ Всего событий произошло: {eventsTriggered,-26} ");
        Console.WriteLine("╚════════════════════════════════════════════════════════╝");
    }
}